sdk/core/Azure.Core/src/Pipeline/Policies/RequestActivityPolicy.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Pipeline.Policies
+{
+    internal class RequestActivityPolicy : HttpPipelinePolicy
+    {
+        public static RequestActivityPolicy Shared { get; } = new RequestActivityPolicy();
+
+        private static readonly DiagnosticListener s_diagnosticSource = new DiagnosticListener("Azure.Pipeline");
+
+        public override Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            return ProcessAsync(message, pipeline, true);
+        }
+
+        public override void Process(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessAsync(message, pipeline, false).EnsureCompleted();
+        }
+
+        private static async Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool isAsync)
+        {
+            if (!s_diagnosticSource.IsEnabled())
+            {
+                if (isAsync)
+                {
+                    await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                }
+                else
+                {
+                    ProcessNext(message, pipeline);
+                }
+
+                return;
+            }
+
+            var activity = new Activity("Azure.Core.Http.Request");
+            activity.AddTag("method", message.Request.Method.Method);
+            activity.AddTag("uri", message.Request.UriBuilder.ToString());
+
+            var diagnosticSourceActivityEnabled = s_diagnosticSource.IsEnabled(activity.OperationName);
+
+            if (diagnosticSourceActivityEnabled)
+            {
+                s_diagnosticSource.StartActivity(activity, null);
+            }
+            else
+            {
+                activity.Start();
+            }
+
+            if (isAsync)
+            {
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+
+            activity.AddTag("status", message.Response.Status.ToString(CultureInfo.InvariantCulture));
```

Comment:
Is it worth checking for common Azure headers like `x-ms-error-code` and adding them in the event of a failure?

Reply:
Maybe :) I didn't go too much into details on what information to put into traces as I'm sure it would be revisited many times. I suspect there would be an integration point to client specific customization to this activity.